### PR TITLE
Reimplement tls-alpn-01 in acme

### PR DESCRIPTION
**Pull request by adferrand**
This PR is the first part of work described in #6724.

It reintroduces the tls-alpn-01 challenge in `acme` module, that was introduced by #5894 and reverted by #6100. The reason it was removed in the past is because some tests showed that with `1.0.2` branch of OpenSSL, the self-signed certificate containing the authorization key is sent to the requester even if the ALPN protocol `acme-tls/1` was not declared as supported by the requester during the TLS handshake.

However recent discussions lead to the conclusion that this behavior was not a security issue, because first it is coherent with the behavior with servers that do not support ALPN at all, and second it cannot make a tls-alpn-01 challenge be validated in this kind of corner case.

On top of the original modifications given by #5894, I merged the code to be up-to-date with our `master`, and fixed tests to match recent evolution about not displaying the `keyAuthorization` in the deserialized JSON form of an ACME challenge.

I also move the logic to verify if ALPN is available on the current system, and so that the tls-alpn-01 challenge can be used, to a dedicated static function `is_available` in `acme.challenge.TLSALPN01`. This function is used in the related tests to skip them, and will be used in the future from Certbot plugins to trigger or not the logic related to tls-alpn-01, depending on the OpenSSL version available to Python.

As a final note about the original error that lead to revert: I could not reproduce any error in `acme` unit tests, from a Ubuntu 17.10 Docker (OpenSSL 1.0.2g) using non-precompiled `pyOpenSSL` package (so linking to the OpenSSL system library). Yet this specific configuration has proven to expose the certificate on an ALPN TLS handshake without `acme-tls/1` been supported. So I will run the test-everything approach here, to see if the error can be found. This PR should also be tested against the test farm before merge.

### CODE REVIEW

**Comment by adferrand** [comment]
Successful build with test-everything approach is available here: https://travis-ci.com/certbot/certbot/builds/105861941

**Comment by dholth** [comment]
The way to detect alpn ahead of time is to use openssl bio (all io goes through python before reaching openssl). Acme will be in the first packet. 

https://github.com/dholth/txsni/blob/f016940f53bffca1c85406c32ad01ce8f7813a0e/txsni/snimap.py#L78

**Comment by adferrand** [comment]
Hello @dholth, your approach is very promising, and would ensure to have the same behavior across all openssl versions, and provide a way to select the right certificate for the standalone server in acme.

However, my lack of experience with OpenSSL makes me fait to write a functional and elegant implementation in `acme` module.

Since not solving is not necessary for implementing tls-alpn, and is not a security breach, I will not try more. 

Do not hesitate however to propose the improvement if you have the time and the motivation, this is for sure a desirable feature. 

@bmw, this PR is ready for review.

**Comment by adferrand** [comment]
@bmw, since you are working on removing totally TLS-SNI-01 with #7479, I wanted to remind you that this PR reintroduce the support of TLS-ALPN-01. I juste merged master to it and the PR is ready for review.

**Comment by bmw** [comment]
Thanks for the reminder.

I'd love to see this merged, but the only benefits I see of bumping the priority of this are:

1. It looks good from a public perspective that ripping out our TLS-SNI code is accompanied by TLS-ALPN but I don't think there's a sudden new need that needs to be filled here as none of our plugins have supported TLS-SNI for quite some time.
2. It'd make it easier for 3rd party plugins to support TLS-ALPN, however, they could still support it now if they wanted.

If I'm missing something, please let me know.

With that said, I'll try to review this if I find the time, but I can't make any promises.

**Comment by adferrand** [comment]
For 1., this PR is in fact the first part of how to introduce tls-alpn-01 support on standalone plugin, that has been asked several times in #6724. It is not a new need however, just an old one that has not been handled so far.

For 2., I do not understand: would having tls-alpn-01 in acme not help third parties plugin to implement this protocol? This is one of the purpose to have a dedicated `acme` library right?

Anyway I do not urge anything here, it was just a reminder if the integration of this PR fits in the timeline, and to keep it ready for merge if so.

About the review, you can keep in mind that most of the code here has been in fact already review and merged once in certbot. I just take back the code, fix the failing tests, and make in compliant with latest ACME protocol specs.

**Comment by bmw** [comment]
> For 2., I do not understand: would having tls-alpn-01 in acme not help third parties plugin to implement this protocol? This is one of the purpose to have a dedicated acme library right?

No I think we agree that it would help 3rd party plugins/clients implement TLS-ALPN. I was just saying that this PR is not a blocker and even Certbot plugins could support the challenge now if they wanted to write similar code to what you had here.

File: acme/acme/challenges.py [hunk 1]
```diff
@@ -8,14 +9,15 @@
...
 from cryptography.hazmat.primitives import hashes  # type: ignore
 import josepy as jose
-import OpenSSL
 import requests
 import six
+from OpenSSL import SSL  # type: ignore # https://github.com/python/typeshed/issues/2052
+from OpenSSL import crypto
 
-from acme import errors
+from acme import _TLSSNI01DeprecationModule
```
[the remark this answers is not in the document]
Comment (adferrand) [reply acme/acme/challenges.py]: Not relevant anymore.

File: acme/acme/challenges.py [hunk 2]
```diff
@@ -616,7 +617,7 @@ def simple_verify(self, chall, domain, account_public_key,
         :param int port: Port used to probe the certificate.
 
 
-        :returns: ``True`` iff client's control of the domain has been
+        :returns: ``True`` if client's control of the domain has been
```
[the remark this answers is not in the document]
Comment (adferrand) [reply acme/acme/challenges.py]: Indeed, iff is not an abreviation I had in my vocabulary, thanks for that piece of information. It is certainly better to be explicit here.

File: acme/acme/challenges.py [hunk 3]
```diff
@@ -511,29 +513,164 @@ def validation(self, account_key, **kwargs):
...
+        return self.response(account_key).gen_cert(
+            key=kwargs.get('cert_key'),
+            domain=kwargs.get('domain'))
+
+    @staticmethod
+    def is_supported():
+        """
+        Check if TLS-ALPN-01 challenge is supported on this machine.
+        This implies that a recent version of OpenSSL is installed (>= 1.0.2),
+        or a recent pre-compiled PyOpenSSL version with embedded OpenSSL library is installed.
```
[the remark this answers is not in the document]
Comment (adferrand) [reply acme/acme/challenges.py]: Corrected.

File: acme/acme/standalone.py [hunk 4]
```diff
@@ -124,6 +131,57 @@ def shutdown_and_server_close(self):
...
+        if len(alpn_protos) == 1 and alpn_protos[0] == self.ACME_TLS_1_PROTOCOL:
+            logger.debug("Agreed on %s ALPN", self.ACME_TLS_1_PROTOCOL)
+            return self.ACME_TLS_1_PROTOCOL
+        # Raising an exception causes openssl to terminate handshake and
+        # send fatal tls alert.
+        logger.debug("Cannot agree on ALPN proto. Got: %s", str(alpn_protos))
+        raise BadALPNProtos("Got: %s" % str(alpn_protos))
+
+
+class BaseRequestHandlerWithLogging(socketserver.BaseRequestHandler):
```
[the remark this answers is not in the document]
Comment (adferrand) [reply acme/acme/standalone.py]: Done.

**Comment by adferrand** [comment]
Added a changelog entry, and fixed behavior with old versions of OpenSSL, validated with the Dockerfile you provided @bmw.

File: acme/acme/standalone.py [hunk 5]
```diff
@@ -120,6 +127,45 @@ def shutdown_and_server_close(self):
...
+        logger.debug("Serving challenge cert for server name %s", server_name)
+        return self.challenge_certs.get(server_name, None)
+
+    def _alpn_selection(self, connection, alpn_protos):
+        """Callback to select alpn protocol."""
+        if len(alpn_protos) == 1 and alpn_protos[0] == self.ACME_TLS_1_PROTOCOL:
+            logger.debug("Agreed on %s ALPN", self.ACME_TLS_1_PROTOCOL)
+            return self.ACME_TLS_1_PROTOCOL
+        logger.debug("Cannot agree on ALPN proto. Got: %s", str(alpn_protos))
+        # Explicitly close the connection now because some old versions of OpenSSL do
```
Comment (bmw) [inline acme/acme/standalone.py]: According to https://www.pyopenssl.org/en/stable/api/ssl.html#OpenSSL.SSL.Context.set_alpn_select_callback, if we want to close the connection we should return an empty bytestring here. I tested this on a few systems and it seemed to always work.

Unless you can find a problem with this approach, I think we should replace the rest of this function with `return b""` and comment explaining what this return value does. If we take this approach, the exception `BadALPNProtos` will no longer be used so I think we should delete it.

**Comment by bmw** [comment]
I just wanted to nudge this PR in case you forgot about it. We're so close to being able to merge it!

**Comment by adferrand** [comment]
So sorry, I missed the mail about your last review! I handle that right now.

File: acme/acme/standalone.py (hunk 5 above)
Comment (adferrand) [reply acme/acme/standalone.py]: Yes you are right. The use of an exception was in the initial PR I took over, but it seems that it was working because recent versions of pyopenssl was properly handling exceptions by returning the empty string, while old ones were not.

Anyway the spec clearly indicates to return an empty string, so better to follow that.

**Comment by sebma** [comment]
@bmw Hi, if this PR is merged, does this mean that TLS-ALPN-01 is now supported and that issue #6724 can be closed ?
